core/trino-main/src/main/java/io/trino/operator/aggregation/partial/ActivePartialAggregationControl.java

```diff
@@ -0,0 +1,109 @@
+ */
+package io.trino.operator.aggregation.partial;
+
+import io.trino.operator.HashAggregationOperator;
+
+/**
+ * Controls whenever partial aggregation is enabled across all {@link HashAggregationOperator}s
+ * for a particular plan node on a single node.
+ * Partial aggregation is disabled once enough rows has been processed ({@link #minNumberOfRowsProcessed})
+ * and the ratio between output(unique) and input rows is too high (> {@link #uniqueRowsRatio}).
+ * <p>
+ * The class is thread safe and objects of this class are used potentially by multiple threads/drivers simultaneously.
+ * Different threads either:
+ * - modify fields via synchronized {@link #onPartialAggregationFlush}.
+ * - read volatile {@link #partialAggregationDisabled} (volatile here gives visibility).
+ */
+public class ActivePartialAggregationControl
+        implements PartialAggregationControl
+{
+    private final long minNumberOfRowsProcessed;
+    private final double uniqueRowsRatio;
+
+    private volatile boolean partialAggregationDisabled;
+    private long totalRowProcessed;
+    private long totalUniqueRowsProduced;
+
+    public ActivePartialAggregationControl(long minNumberOfRowsProcessedToDisable, double uniqueRowsRatio)
+    {
+        this.minNumberOfRowsProcessed = minNumberOfRowsProcessedToDisable;
+        this.uniqueRowsRatio = uniqueRowsRatio;
+    }
+
+    public synchronized void onPartialAggregationFlush(long rowsProcessed, long uniqueRowsProduced)
+    {
+        totalRowProcessed += rowsProcessed;
+        totalUniqueRowsProduced += uniqueRowsProduced;
+        if (shouldDisablePartialAggregation()) {
+            partialAggregationDisabled = true;
+        }
+    }
+
+    private boolean shouldDisablePartialAggregation()
+    {
+        return totalRowProcessed >= minNumberOfRowsProcessed
+                && ((double) totalUniqueRowsProduced / totalRowProcessed) > uniqueRowsRatio;
+    }
+
+    @Override
+    public PartialAggregationTracker newTracker()
+    {
+        return new ActivePartialAggregationTracker();
+    }
+
+    /**
+     * Tracks number of input rows processed and unique groups produced
+     * per single Partial Aggregation hash table (i.e. per flush).
+     */
+    public class ActivePartialAggregationTracker
+            implements PartialAggregationTracker
+    {
+        private long numberOfRowsProcessed;
+        private long numberOfUniqueRowsProduced;
+
+        @Override
+        public boolean isPartialAggregationDisabled()
+        {
+            return partialAggregationDisabled;
+        }
+
+        @Override
+        public void addRowsProcessed(int positionCount)
+        {
+            numberOfRowsProcessed += positionCount;
+        }
+
+        @Override
+        public void addUniqueRowsProduced(int positionCount)
+        {
+            numberOfUniqueRowsProduced += positionCount;
+        }
+
+        @Override
+        public void onFlush()
+        {
+            if (!isPartialAggregationDisabled()) {
```

Comment:
Just move this logic to `onPartialAggregationFlush` (make tracker dumb). `onFlush` is not frequent